codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -317,125 +314,60 @@ private Http2Stream requireStream(int streamId) {
      * </p>
      */
     private final class FlowControlledData extends FlowControlledBase {
-        private ByteBuf data;
-        private int size;
+
+        private final CoalescingBufferQueue queue;
 
         private FlowControlledData(ChannelHandlerContext ctx, Http2Stream stream, ByteBuf buf, int padding,
                                     boolean endOfStream, ChannelPromise promise) {
             super(ctx, stream, padding, endOfStream, promise);
-            this.data = buf;
-            size = data.readableBytes() + padding;
+            queue = new CoalescingBufferQueue(ctx.channel());
+            queue.add(buf, promise);
         }
 
         @Override
         public int size() {
-            return size;
+            return queue.readableBytes() + padding;
         }
 
         @Override
         public void error(Throwable cause) {
-            ReferenceCountUtil.safeRelease(data);
+            queue.releaseAndFailAll(cause);
             lifecycleManager.onException(ctx, cause);
-            data = null;
-            size = 0;
             promise.tryFailure(cause);
         }
 
         @Override
         public void write(int allowedBytes) {
-            int bytesWritten = 0;
-            if (data == null || (allowedBytes == 0 && size != 0)) {
-                // No point writing an empty DATA frame, wait for a bigger allowance.
+            if (!endOfStream && (queue.readableBytes() == 0 || allowedBytes == 0)) {
```

Comment:
Nvm...I see the additional `endOfStream` restriction is now being imposed...which makes sense.